### PR TITLE
Alt drag to duplicate

### PR DESCRIPTION
**Pull request by fakob**
* Added option to alt+drag to duplicate
* Made pasteNodes undoable

https://github.com/fakob/plug-and-play/assets/4619772/6821de3d-b29f-461e-b53a-f32a968cdc23

### CODE REVIEW

File: src/classes/GraphClass.ts [hunk 1]
```diff
@@ -761,68 +766,82 @@ export default class PPGraph {
     this.allowExecution = true;
   }
 
-  async duplicateSelection(): Promise<PPNode[]> {
+  async duplicateSelection(pastePos: TPastePos = undefined): Promise<PPNode[]> {
```
Comment (magnificus) [inline src/classes/GraphClass.ts:769]: can we avoid the undefined here? Maybe by making pastepos a mandatory parameter or add existing default value?

File: src/classes/NodeClass.ts [hunk 2]
```diff
@@ -987,14 +987,28 @@ ${Math.round(this._bounds.minX)}, ${Math.round(
...
+      const altKey = event.altKey;
 
       // select node if the shiftKey is pressed
       // or the node is not yet selected
       if (shiftKey || !this.selected) {
         PPGraph.currentGraph.selection.selectNodes([this], shiftKey, true);
         this.onSpecificallySelected();
       }
-      PPGraph.currentGraph.selection.startDragAction(event);
+      if (altKey) {
```
Comment (magnificus) [inline src/classes/NodeClass.ts:998]: getting a lot of ifs in here becomes a bit hard to digest

File: src/classes/GraphClass.ts [hunk 3]
```diff
@@ -761,68 +766,82 @@ export default class PPGraph {
...
     data: SerializedSelection,
-    pasteTo?: {
-      x: number;
-      y: number;
-    }
+    pastePos?: TPastePos
   ): Promise<PPNode[]> {
     const newNodes: PPNode[] = [];
     const mappingOfOldAndNewNodes: { [key: string]: PPNode } = {};
+    const arrayOfRandomIds = createArrayOfRandomIds(data.nodes.length);
```
Comment (magnificus) [inline src/classes/GraphClass.ts:784]: cant you just create the random ids when you create the node further down? Can make it less stateful that way, you already have the list of all nodes so you can get their IDS from there for the undo action.

File: src/utils/utils.ts [hunk 4]
```diff
@@ -840,3 +841,11 @@ export const updateDataIfDefault = (
     node.executeOptimizedChain();
   }
 };
+
+export const createArrayOfRandomIds = (length: number): string[] => {
```
Comment (magnificus) [inline src/utils/utils.ts:845]: dont think we should need this function it seems redundant

Comment (magnificus) [reply src/utils/utils.ts:845]: Also if we need it we can use

Array.from({ length },() => hri.random())

**Review by magnificus — APPROVED** [review]
approved but added some comments

File: src/classes/GraphClass.ts (hunk 1 above)
Comment (fakob) [reply src/classes/GraphClass.ts:769]: This is a similar issue as commented in the other PR. 
* When defined we are doing a general offset based on the pastePos
* When undefined we are doing an individual positioning based on the original nodes position. We are solving the issue downstream when looping through the individual nodes.

File: src/classes/NodeClass.ts (hunk 2 above)
Comment (fakob) [reply src/classes/NodeClass.ts:998]: I agree, but I am not sure how to solve it. If I move to use a switch I would have to duplicate code as they are not mutually exclusive. What could be another option?

File: src/classes/GraphClass.ts (hunk 3 above)
Comment (fakob) [reply src/classes/GraphClass.ts:784]: In the undo action I have to delete all newly created nodes. They are deleted using/identified by the random IDs I create here. The IDs from the original nodes would not help me in the undo action as I am not touching them.
Looking at other undoable actions we had always created the Ids for the newly created nodes before. Is there another way to keep track of them?

File: src/utils/utils.ts (hunk 4 above)
Comment (fakob) [reply src/utils/utils.ts:845]: Thanks for the short version. I was not aware of the Array.from usage.

File: src/classes/GraphClass.ts (hunk 1 above)
Comment (magnificus) [reply src/classes/GraphClass.ts:765]: what about another function without parameters that calls the second one?

Comment (fakob) [reply src/classes/GraphClass.ts:765]: I am not sure I fully understand. Like this?
```
  duplicateSelectionWithoutPosition() {
    this.duplicateSelection();
  }

  async duplicateSelection(pastePos: TPastePos = undefined): Promise<PPNode[]> {
    const serializeSelection = this.serializeSelection();
    const pastedNodes = await this.action_pasteNodes(
      serializeSelection,
      pastePos
    );
    return pastedNodes;
  }
```
I would still have to pass in undefined? I will merge and we can discuss this further when you are back.

**Pull request merged**